File: partiql-plan/src/main/java/org/partiql/plan/rex/RexError.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;
import org.partiql.spi.types.PType;

import java.util.List;

/**
 * This represents scenarios in which certain operations are statically known to fail in strict mode but return missing
 * in permissive mode.
 */
public abstract class RexError extends RexBase {

    /**
     * @return new RexError instance
     */
    @NotNull
    public static RexError create() {
        return new Impl();
    }

    @NotNull
    @Override
    protected RexType type() {
        // TODO SHOULD BE UNKNOWN
        return RexType.of(PType.dynamic());
    }

    @NotNull
    @Override
    protected List<Operand> operands() {
        return List.of();
    }

    @Override
    public <R, C> R accept(OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitError(this, ctx);
    }

    private static class Impl extends RexError {
        // empty
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rex/RexError.kt
package org.partiql.plan.rex

/**
 * This represents scenarios in which certain operations are statically known to fail in strict mode but return missing
 * in permissive mode.
 */
import org.partiql.plan.Visitor

public interface RexError : Rex {

    override fun getType(): RexType = RexType.dynamic()

    override fun getChildren(): Collection<Rex> = emptyList()

    override fun <R, C> accept(visitor: Visitor<R, C>, ctx: C): R = visitor.visitError(this, ctx)
}

internal class RexErrorImpl : RexError
